File: app/src/main/java/org/neidhardt/dynamicsoundboard/navigationdrawer/views/NavigationDrawerListPresenter.java
package org.neidhardt.dynamicsoundboard.navigationdrawer.views;

import android.view.Menu;
import android.view.MenuItem;
import org.neidhardt.dynamicsoundboard.navigationdrawer.events.ActionModeChangeRequestedEvent;
import org.neidhardt.dynamicsoundboard.presenter.BaseViewPresenter;

/**
 * File created by eric.neidhardt on 26.05.2015.
 */
public abstract class NavigationDrawerListPresenter<T extends NavigationDrawerList>
		extends
			BaseViewPresenter<T>
		implements
			android.support.v7.view.ActionMode.Callback
{
	private static final String TAG = NavigationDrawerListPresenter.class.getName();

	private boolean isInSelectionMode;

	protected void onItemSelectedForDeletion()
	{
		this.getEventBus().post(new ActionModeChangeRequestedEvent(this, ActionModeChangeRequestedEvent.REQUEST.INVALIDATE));
	}

	public void prepareItemDeletion()
	{
		this.getEventBus().post(new ActionModeChangeRequestedEvent(this, ActionModeChangeRequestedEvent.REQUEST.START));
	}

	protected void onSelectedItemsDeleted()
	{
		this.getEventBus().post(new ActionModeChangeRequestedEvent(this, ActionModeChangeRequestedEvent.REQUEST.STOP));
	}

	@Override
	public boolean onCreateActionMode(android.support.v7.view.ActionMode actionMode, Menu menu)
	{
		if (this.getView() == null)
			throw new NullPointerException(TAG + ".onCreateActionMode failed, supplied view is null ");

		this.isInSelectionMode = true;
		return true;
	}

	@Override
	public boolean onPrepareActionMode(android.support.v7.view.ActionMode actionMode, Menu menu)
	{
		if (this.getView() == null)
			throw new NullPointerException(TAG + ".onPrepareActionMode failed, supplied view is null ");

		actionMode.setTitle(this.getView().getActionModeTitle());

		int count = this.getNumberOfItemsSelectedForDeletion();
		String countString = Integer.toString(count);
		if (countString.length() == 1)
			countString = " " + countString;
		countString = countString + "/" + this.getView().getItemCount();

		actionMode.setSubtitle(countString);
		return true;
	}

	@Override
	public boolean onActionItemClicked(android.support.v7.view.ActionMode actionMode, MenuItem menuItem)
	{
		return false;
	}

	@Override
	public void onDestroyActionMode(android.support.v7.view.ActionMode actionMode)
	{
		this.isInSelectionMode = false;
		this.deselectAllItemsSelectedForDeletion();

		this.getEventBus().post(new ActionModeChangeRequestedEvent(this, ActionModeChangeRequestedEvent.REQUEST.STOPPED));
	}

	public abstract void deleteSelectedItems();

	protected abstract int getNumberOfItemsSelectedForDeletion();

	protected abstract void deselectAllItemsSelectedForDeletion();

	public boolean isInSelectionMode()
	{
		return isInSelectionMode;
	}

	public void setIsInSelectionMode(boolean isInSelectionMode)
	{
		this.isInSelectionMode = isInSelectionMode;
	}
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/navigationdrawer/views/NavigationDrawerListPresenter.kt
package org.neidhardt.dynamicsoundboard.navigationdrawer.views

import android.view.Menu
import android.view.MenuItem
import org.neidhardt.dynamicsoundboard.navigationdrawer.events.ActionModeChangeRequestedEvent
import org.neidhardt.dynamicsoundboard.presenter.BaseViewPresenter

/**
 * File created by eric.neidhardt on 26.05.2015.
 */
abstract class NavigationDrawerListPresenter<T: NavigationDrawerList> : BaseViewPresenter<T>(), android.support.v7.view.ActionMode.Callback
{
	private val TAG = javaClass.name

	var isInSelectionMode: Boolean = false

	protected fun onItemSelectedForDeletion() = this.eventBus.post(ActionModeChangeRequestedEvent(this, ActionModeChangeRequestedEvent.REQUEST.INVALIDATE))

	protected fun onSelectedItemsDeleted() = this.eventBus.post(ActionModeChangeRequestedEvent(this, ActionModeChangeRequestedEvent.REQUEST.STOP))

	fun prepareItemDeletion() = this.eventBus.post(ActionModeChangeRequestedEvent(this, ActionModeChangeRequestedEvent.REQUEST.START))

	override fun onCreateActionMode(actionMode: android.support.v7.view.ActionMode, menu: Menu): Boolean
	{
		if (this.view == null)
			throw NullPointerException(TAG + ".onCreateActionMode failed, supplied view is null ")

		this.isInSelectionMode = true
		return true
	}

	override fun onPrepareActionMode(actionMode: android.support.v7.view.ActionMode, menu: Menu): Boolean
	{
		if (this.view == null)
			throw NullPointerException(TAG + ".onPrepareActionMode failed, supplied view is null ")

		actionMode.setTitle(this.view.actionModeTitle)

		val count = this.numberOfItemsSelectedForDeletion
		var countString = Integer.toString(count)
		if (countString.length() == 1)
			countString = " " + countString
		countString = countString + "/" + this.view.itemCount

		actionMode.subtitle = countString
		return true
	}

	override fun onActionItemClicked(actionMode: android.support.v7.view.ActionMode, menuItem: MenuItem): Boolean = false

	override fun onDestroyActionMode(actionMode: android.support.v7.view.ActionMode)
	{
		this.isInSelectionMode = false
		this.deselectAllItemsSelectedForDeletion()

		this.eventBus.post(ActionModeChangeRequestedEvent(this, ActionModeChangeRequestedEvent.REQUEST.STOPPED))
	}

	abstract fun deleteSelectedItems()

	protected abstract val numberOfItemsSelectedForDeletion: Int

	protected abstract fun deselectAllItemsSelectedForDeletion()
}
